interfaces/acados_template/acados_template/mpc_utils.py

```diff
@@ -28,37 +28,355 @@
+        """
+        Evaluates the cost and constraints at a given stage of the OCP. For a closed-loop evaluation the stage is
+        typically 0.
+        @param x: state vector
+        @param u: control input vector
+        @param step: stage index (0 <= stage < N)
+        @return: dictionary with the following keys:
+            - 'cost': total cost of the transition
+            - 'cost_without_slacks': total cost of transition without slack penalties
+            - 'cost_slacks': total cost of slack penalties
+            - 'violation_soft_constraints': individual violation of soft constraints (equal to slacks)
+            - 'violation_hard_constraints': individual violation of hard constraints
+        """
+
+        cost_fun_args = [x, u, self.__parameter_values[:,step], self.__p_global_values]
+
+        # evaluate cost
+        cost_without_slacks = self.cost_fun(*cost_fun_args).full() * self.time_steps[step]
+
+        # evaluate constraints
+        lower_violation, upper_violation, lower_slack, upper_slack = (
+            self.constraint_function(x, u,
+                                     self.__parameter_values[:,step],
+                                     self.__p_global_values))
+        violation_hard_constraints = np.concatenate(
+            (lower_violation[self.nonslacked_indices], upper_violation[self.nonslacked_indices]))
+
+        # evaluate cost of soft constraints
+        lower_slack_cost, upper_slack_cost = np.array([0.]), np.array([0.])
+
+        if self.ocp.cost.Zl.size > 0:
+            lower_slack_cost += 0.5 * self.ocp.cost.Zl @ (lower_slack.full()  * lower_slack.full())
+        if self.ocp.cost.zl.size > 0:
+            lower_slack_cost += self.ocp.cost.zl @ lower_slack.full()
+
+        if self.ocp.cost.Zu.size > 0:
+            upper_slack_cost += 0.5 * self.ocp.cost.Zu @ (upper_slack.full() * upper_slack.full() )
+        if self.ocp.cost.zu.size > 0:
+            upper_slack_cost += self.ocp.cost.zu @ upper_slack.full()
+
+        slack_cost = (lower_slack_cost + upper_slack_cost) * self.time_steps[step]
+
+        if len(slack_cost) == 0:
+            cost = cost_without_slacks
+        else:
+            cost = cost_without_slacks + slack_cost
+
+        # evaluate sum
+        result = {
+            'cost': array_to_float(cost),
+            'cost_without_slacks': array_to_float(cost_without_slacks),
+            'cost_slacks': array_to_float(slack_cost),
+            'violation_soft_constraints': np.concatenate((lower_slack.full(), upper_slack.full())),
+            'violation_hard_constraints': violation_hard_constraints,
+        }
+        return result
+
+    def evaluate_ocp_cost(self, acados_ocp_iterate: AcadosOcpIterate):
+        """
+        Evaluates the cost of a whole OCP trajectory, as evaluated inside acados.
+        """
+        acados_ocp_iterate = deepcopy(acados_ocp_iterate)
+        acados_ocp_iterate.flatten()
+        x_traj = np.array(acados_ocp_iterate.x_traj)
```

Comment:
why do you need to cast to array here? If you directly iterate over the list then also MOCP formulations would be supported by the evaluator